libbeat/processors/rate_limit/algorithm/token_bucket.go

```diff
@@ -149,15 +156,15 @@ func (b *bucket) replenish(rate Rate, clock clock.Clock) {
 
 func (t *tokenBucket) runGC() {
 	// Don't run GC if thresholds haven't been crossed.
-	if (t.gc.metrics.NumBuckets < t.gc.thresholds.NumBuckets) &&
-		(t.gc.metrics.NumCalls < t.gc.thresholds.NumCalls) {
+	if (t.gc.metrics.numBuckets.Load() < t.gc.thresholds.NumBuckets) &&
+		(t.gc.metrics.numCalls.Load() < t.gc.thresholds.NumCalls) {
 		return
 	}
 
```

Comment:
By using atomic we won't have race conditions on the counting, but I think we still need a non-blocking mutex here to avoid having more than one GC running at the same time. If multiple goroutines call `runGC()` at the same time when one of the thresholds have been reached, all of them will start doing the job.

Reply:
Implemented in 845b17b388683dcd8dcc0ec42379f2ac2e88a0e1.